File: go/0040-combination-sum-ii.go
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	ans := make([][]int, 0)
	curr := make([]int, 0)
	sort.Ints(candidates)
	var backtrack func(idx int, currSum int, curr []int)
	backtrack = func(idx int, currSum int, curr []int) {
		if currSum == target {
			ans = append(ans, append([]int{}, curr...))
			return
		}
		if currSum > target {
			return
		}
		for i := idx; i < len(candidates); i++ {
			if i > idx && candidates[i] == candidates[i-1] {
				continue
			}
			curr = append(curr, candidates[i])
			backtrack(i+1, currSum+candidates[i], curr)
			curr = curr[:len(curr)-1]
		}

	}
	backtrack(0, 0, curr)
	return ans
}

File: ruby/0040-combination-sum-ii.rb
def combination_sum2(candidates, target)
    @candidates = candidates
    @candidates.sort!
    @result = []

    def backtrack (cur,pos,target)
        @result.append(cur.dup()) if target == 0
        return if target <=0

        prev = -1

        (pos..@candidates.length-1).each do |i|
            next if @candidates[i] == prev

            cur.append(@candidates[i])

            backtrack(cur,i+1,target-@candidates[i])

            cur.pop()

            prev = @candidates[i]

        end

    end
    backtrack([],0,target)

    return @result

    
end